builtins/web/form-data/form-data-encoder.cpp

```diff
@@ -0,0 +1,714 @@
+  auto entry = stream.entries->begin()[chunk_idx_];
+
+  if (entry.value.isString()) {
+    RootedValue value_val(cx, entry.value);
+    auto maybe_normalized = normalize_newlines(cx, value_val);
+    if (!maybe_normalized) {
+      return false;
+    }
+
+    auto normalized = maybe_normalized.value();
+    write_and_cache_remainder(stream, normalized.begin(), normalized.end());
+  } else {
+    MOZ_ASSERT(File::is_instance(entry.value));
+    RootedObject obj(cx, &entry.value.toObject());
+
+    auto blob = Blob::blob(obj);
+    auto blobsz = blob->length();
+    auto written = stream.write(blob->begin(), blob->end());
+    MOZ_ASSERT(written <= blobsz);
+    file_leftovers_ = blobsz - written;
+  }
+
+  return true;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.1 - writes `crlf`
+bool MultipartFormDataImpl::handle_entry_footer(JSContext *cx, StreamContext &stream) {
+  auto footer = fmt::memory_buffer();
+  fmt::format_to(std::back_inserter(footer), "\r\n");
+
+  write_and_cache_remainder(stream, footer.begin(), footer.end());
+  chunk_idx_ += 1;
+
+  MOZ_ASSERT(chunk_idx_ <= stream.entries->length());
+  return true;
+}
+
+// https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.1
+//
+// The boundary delimiter line following the last body part is a distinguished delimiter that
+// indicates that no further body parts will follow.  Such a delimiter line is identical to
+// the previous delimiter lines, with the addition of two more hyphens after the boundary
+// parameter value.
+bool MultipartFormDataImpl::handle_close(JSContext *cx, StreamContext &stream) {
+  auto footer = fmt::memory_buffer();
+  fmt::format_to(std::back_inserter(footer), "--{}--", boundary_);
+
+  write_and_cache_remainder(stream, footer.begin(), footer.end());
+  return true;
+}
+
+bool MultipartFormDataImpl::read_next(JSContext *cx, StreamContext &stream) {
+  maybe_drain_leftovers(cx, stream);
+  if (is_draining()) {
+    return true;
+  }
+
+  state_ = next_state(stream);
+
+  switch (state_) {
+  case State::EntryHeader: {
+    return handle_entry_header(cx, stream);
+  }
+  case State::EntryBody: {
+    return handle_entry_body(cx, stream);
+  }
+  case State::EntryFooter: {
+    return handle_entry_footer(cx, stream);
+  }
+  case State::Close: {
+    return handle_close(cx, stream);
+  }
+  case State::Done: {
+    stream.done = true;
+    return true;
+  }
+  default:
+    MOZ_ASSERT_UNREACHABLE("Invalid state");
+    return false;
+  }
+}
+
+// Computes the total size (in bytes) of the encoded multipart/form-data stream.
+//
+// Returns `std::nullopt` if any string conversion fails. This function simulates
+// the multipart/form-data encoding process without actually writing to a buffer.
+// Instead, it accumulates the total size of each encoding step.
+std::optional<size_t> MultipartFormDataImpl::query_length(JSContext* cx, const EntryList *entries) {
```

Comment:
Not too important, but maybe use `mozilla::Result` for this instead of `std::optional`?